File: hud/view/device_dialog.py
from PySide6.QtCore import Signal, Qt
from PySide6.QtGui import QPainter, QBrush, QColor
from PySide6.QtWidgets import QDialog, QListWidget, QHBoxLayout, QVBoxLayout, QListWidgetItem

from hud.configuration.config import Config
from hud.view.primitives.labels import ButtonLabel
from hud.view.primitives.theme_switch import with_switchable_theme


@with_switchable_theme
class DeviceDialog(QDialog):
    selectedDevice = None
    selectDeviceSignal = Signal(object)

    def __init__(self, app_config: Config, parent=None):
        super().__init__(parent)
        self.app_config = app_config

        self.selectedDevice = None

        self.listWidget = None
        self.closeLabel = None

        self.layout = None
        self.hLayout = None

        theme = self.app_config.hud_layout.theme

        self.listWidget = QListWidget(self)
        self.listWidget.itemClicked.connect(self.selectItem)
        self.listWidget.itemDoubleClicked.connect(self.confirmSelection)
        self.listWidget.setStyleSheet(theme.color_scheme_stylesheet)

        self.closeLabel = ButtonLabel(text="Ok", theme=theme)
        self.closeLabel.clicked.connect(self.onLabelClicked)

        self.hLayout = QHBoxLayout()
        self.hLayout.addStretch()
        self.hLayout.addWidget(self.closeLabel)
        self.hLayout.addStretch()

        self.layout = QVBoxLayout(self)
        self.layout.addWidget(self.listWidget)
        self.layout.addLayout(self.hLayout)

        self.setWindowFlags(Qt.WindowType.FramelessWindowHint | Qt.WindowType.Dialog)
        self.setAttribute(Qt.WidgetAttribute.WA_TranslucentBackground)

    def onLabelClicked(self):
        self.selectDeviceSignal.emit(self.selectedDevice)
        self.close()

    def closeEvent(self, event):
        event.accept()

    def showDevice(self, device):
        item = QListWidgetItem(device.name)
        item.setData(Qt.ItemDataRole.UserRole, device)
        if self.listWidget.findItems(device.name, Qt.MatchFlag.MatchExactly):
            return
        self.listWidget.addItem(item)

    def selectItem(self, item):
        self.selectedDevice = item.data(Qt.ItemDataRole.UserRole)

    def confirmSelection(self, item):
        device = item.data(Qt.ItemDataRole.UserRole)

        self.selectDeviceSignal.emit(device)
        self.accept()

    def paintEvent(self, event):
        painter = QPainter(self)
        painter.setOpacity(0.75)  # Set the opacity
        painter.setBrush(QBrush(QColor(*self.app_config.hud_layout.theme.background_colour)))
        painter.drawRect(self.rect())

    async def show_async(self):
        # self.exec()
        self.show()
        self.activateWindow()

File: src/DeviceDialog.cpp
#include "DeviceDialog.h"

#include <iostream>
#include <qboxlayout.h>
#include <qevent.h>
#include <QListWidget>
#include <QPainter>

#include "Constants.h"
#include "Events.h"
#include "Labels.h"
#include "StyleSheets.h"

DeviceDialog::DeviceDialog(
    const std::vector<std::shared_ptr<Device> > &data,
    const std::shared_ptr<ControllerHandler> &handler, QWidget *parent
): QDialog(parent),
   handler(handler), listWidget(std::make_shared<QListWidget>(this)) {
    connect(listWidget.get(), &QListWidget::itemClicked, this, &DeviceDialog::itemSelected);
    connect(listWidget.get(), &QListWidget::itemDoubleClicked, this, &DeviceDialog::itemConfirmed);

    for (const auto &device: data) {
        renderDevice(device);
    }
    const auto closeLabel = new ButtonLabel(Constants::Buttons::OK, true, this);
    connect(closeLabel, &ButtonLabel::clicked, this, [this] { this->close(); });

    const auto hLayout = new QHBoxLayout();
    hLayout->addStretch();
    hLayout->addWidget(closeLabel);
    hLayout->addStretch();

    const auto layout = new QVBoxLayout(this);
    layout->addWidget(listWidget.get());
    layout->addLayout(hLayout);

    setLayout(layout);
    setStyleSheet((StyleSheets::THEME_DARK + StyleSheets::SCALE_MEDIUM).data());
    setWindowFlags(Qt::WindowType::FramelessWindowHint | Qt::WindowType::Dialog);
    setAttribute(Qt::WidgetAttribute::WA_TranslucentBackground);
}

void DeviceDialog::itemConfirmed(const QListWidgetItem *item) {
    itemSelected(item);
    close();
}

void DeviceDialog::renderDevice(const std::shared_ptr<Device> &device) const {
    const auto deviceName = QString::fromStdString(device->name.value);

    const auto item = new QListWidgetItem(deviceName);
    item->setData(Qt::ItemDataRole::UserRole, QVariant::fromValue(device));

    const auto existingItems = listWidget->findItems(deviceName, Qt::MatchFlag::MatchExactly);

    if (existingItems.empty()) {
        listWidget->addItem(item);
        return;
    }

    const auto itemToReplace = existingItems.first();
    const auto row = listWidget->row(itemToReplace);
    listWidget->insertItem(row, item);
    listWidget->takeItem(row + 1);
}

void DeviceDialog::itemSelected(const QListWidgetItem *item) {
    std::cout << "DeviceDialog::itemSelected" << std::endl;

    auto data = item->data(Qt::ItemDataRole::UserRole);
    if (data.isNull()) {
        std::cout << "  No data found" << std::endl;
    }

    const auto device = get<std::shared_ptr<Device> >(data);

    std::cout << "  Selected device: " << device->name.value << std::endl;
    selectedItem = device;
}

void DeviceDialog::closeEvent(QCloseEvent *event) {
    std::cout << "DeviceDialog::closeEvent" << std::endl;

    if (selectedItem) {
        std::cout << "  DeviceDialog::closeEvent: emitting deviceSelected" << selectedItem << std::endl;
    }

    handler->next(Constants::Commands::CONNECT);
    update();
    event->accept();
}

bool DeviceDialog::event(QEvent *event) {
    if (event->type() > QEvent::MaxUser or event->type() < QEvent::User) {
        return QDialog::event(event);
    }

    const auto deviceEvent = dynamic_cast<DeviceDiscoveredEvent *>(event);
    if (not deviceEvent) {
        return QDialog::event(event);
    }

    const std::shared_ptr<Device> device = deviceEvent->getDevice();
    renderDevice(device);
    return true;
}

void DeviceDialog::paintEvent(QPaintEvent *event) {
    const auto painter = std::make_unique<QPainter>(this);

    painter->setOpacity(0.75);
    painter->setBrush(QBrush(QColor(200, 200, 200, 128)));
    painter->drawRect(this->rect());
}
